adapt_common/norms.py

```diff
@@ -0,0 +1,163 @@
+"""Module containing overloaded versions of Firedrake's norm and errornorm functions."""
+
+import firedrake as fd
+import ufl
+from firedrake.petsc import PETSc
+
+from adapt_common.utility import cofunction2function
+
+__all__ = ["errornorm", "norm"]
+
+
+@PETSc.Log.EventDecorator()
+def norm(v, norm_type="L2", condition=None, boundary=False):
+    r"""Overload :func:`fd.norms.norm` to allow for :math:`\ell^p` norms.
+
+    Currently supported ``norm_type`` options:
+    * ``'l1'``
+    * ``'l2'``
+    * ``'linf'``
+    * ``'L2'``
+    * ``'Linf'``
+    * ``'H1'``
+    * ``'Hdiv'``
+    * ``'Hcurl'``
+    * or any ``'Lp'`` with :math:`p >= 1`.
+
+    Note that this version is case sensitive, i.e. ``'l2'`` and ``'L2'`` will give
+    different results in general.
+
+    :arg v: the function to take the norm of
+    :type v: :class:`fd.function.Function` or
+        :class:`fd.cofunction.Cofunction`
+    :kwarg norm_type: the type of norm to use
+    :type norm_type: :class:`str`
+    :kwarg condition: a UFL condition for specifying a subdomain to compute the norm
+        over
+    :kwarg boundary: if ``True``, the norm is computed over the domain boundary
+    :type boundary: :class:`bool`
+    :returns: the norm value
+    :rtype: :class:`float`
+    """
+    if isinstance(v, fd.Cofunction):
+        v = cofunction2function(v)
+    condition = condition or fd.Constant(1.0)
+    norm_codes = {"l1": 0, "l2": 2, "linf": 3}
+    p = 2
+    if norm_type in norm_codes or norm_type == "Linf":
+        if boundary:
+            not_impl_err = "lp errors on the boundary not yet implemented."
+            raise NotImplementedError(not_impl_err)
+        v.interpolate(condition * v)
+        with v.dat.vec_ro as vv:
+            if norm_type == "Linf":
+                return vv.max()[1]
```

Comment:
Does this need an abs()? I presume "Linf" is in fact the same as "linf" ?

Reply:
Good point. Dropped `Linf` in 979cd12.